README.md
# Medle

MedLé is a platform for french hospitals to declare their medico-legal activity.

## 👔 Install

First, install git, yarn, docker, docker-compose with [brew](https://brew.sh/) on Mac OS.

Then, run the containers with `docker-compose`.

`docker-compose up --build -d`

Then, the DB is exposed on port 5434 and the app is accessible on port 80.

Connect to the DB via a Postgresql client. For start, there a user with the name `user` and the password `password`.

`docker exec -it medle_db_1 psql -U user`

Create the medle database and the user medle with the password of your choice.

```sql
create database medle;
create user medle with encrypted password 'jJFWsfW5ePbN7J';
grant all privileges on database medle to medle
```

Then create/modify an `.env` file in the root of the project (see `.env.dev` as a reference).

For DATABASE_URL, be sure to use the matching password you have just created before.

```js
NODE_ENV=development

API_URL=http://localhost/api

# for container app usage
POSTGRES_SSL=false
DATABASE_URL=psql://medle:jJFWsfW5ePbN7J@db:5432/medle

# JWT
JWT_SECRET=NEGLaRS3n9JHuY

# Test variables
#TEST_CURRENT_DATE=10/09/2019
```

Then rerun the container app, to force usage of this `.env` file.
`docker-compose up --build -d app`

This is supposed to work now!


## 🏗️ Development usage

For development purpose, it is more handy to use `yarn dev` to benefit from the hot reload offered by Next.
Just copy the file `.env.dev` given as reference and named it as `.env` to be used by dotenv.

With this configuration, the API will run on `localhost:3000` and the database will be a Docker container running on your machine.

So you only have to run the db container (and not the app): `docker-compose up --build -d db`

Now run `yarn` to install the dependencies, then `yarn dev`.

At the end, the app is running at http://localhost:3000.

*Nota bene*

Be careful. If you use the app from the container (as we advise you), it is considered inside the docker network.
So you have to use an DATABASE_URL with __5432__ port, like `DATABASE_URL=psql://medle:bHrdeGk63cHQa7@db:5432/medle`.

On the other hand, when you use `yarn dev`, the node process is on localhost and is therefore outside the docker network.
In this case, you need to use a DATABASE_URL with __5434__ port, like `DATABASE_URL=psql://medle:bHrdeGk63cHQa7@localhost:5434/medle`.

Don't forgete also that in Docker, the db is not in non SSL mode.
So the `POSTGRES_SSL` must be set to false.

With a managed Postgres, `POSTGRES_SSL` must be set to true.

## 🌱 Migration and seeds

The database structure is able to evolve thanks to Knex.js migrations.

To initiate a migration, the easiest way is to use `migrate:make` script in package.json.

```shell
yarn knex migrate:make my-name-of-migration
```

Modify it accordingly to the business needs.

To apply it, use `migrate:latest` script in package.json.

In development mode
`yarn migrate:latest`

In production mode
`sudo docker-compose exec app yarn migrate:latest`

So on another platform like production, the pattern is:

```sh
git pull
sudo docker-compose up --build -d
sudo docker-compose exec app yarn migrate:latest
```

If you're not happy with the migration done, you can rollback with the script `migrate:rollback` in package.json.

`sudo docker-compose exec app yarn migrate:rollback`

On the development platform (local or staging environment), you may need to populate table in JS (you may yet do it directly with SQL client too).

Make a new seed file, in src/knex/seeds/development or src/knex/seeds/staging, then:

In development mode, for applying the development seeds:
`yarn seed:run:dev`

In production mode, for applying the staging seeds:
`sudo docker-compose exec app yarn seed:run:staging`

Never, never, never <strike>give up</strike> apply this seeds in real production environment (under penalty of fetching a Postgres backup in Azure 😭).

Knex migration tips: https://medium.com/@j3y/beyond-basic-knex-js-database-migrations-22263b0fcd7c

### Migrations organization

For migraton involving a table (creation, modificaton including index management, trigger management, etc.), this should be done in the directory src/knex/migrations like the file 20200317163657_create_indexes.js for example.

For migrations involving a stored function or procedure, a view, materialized or not, this should be done in directory src/knex/versions, with a version number, with the master migration file mentioning it.

For example, see this migration file named `20200312161718_create_function_avg_acts.js`:

```js
const v1 = require("../versions/functions/avg_acts/v1")

exports.up = async function(knex) {
   await knex.raw(v1.up)
}

exports.down = async function(knex) {
   await knex.raw(v1.down)
}
```

This way allows us to see the entire history of things like functions or views, which can be totally removed and recreated.

## 🧪 How to test

In development mode:
1. `yarn dev`
2. `yarn seed:run:staging` (will remove all data on env.DATABASE_URL and reset with the default data)
3. `yarn test` (will use env.API_URL for tests involving API endpoints)

In staging mode:
1. `sudo docker-compose up --build -d app`
2. `sudo docker-compose exec app yarn seed:run:staging` (will remove all data on env.DATABASE_URL and reset with the default data)
3. `sudo docker-compose exec app yarn test` (will use env.API_URL for tests involving API endpoints)

## 🗂️ Main src directories

- clients: the API clients for the frontend
- components: React component
- faq: the mdx file for dynamically populating the FAQ page
- knex: all Knex database backend stuff
- models: the transformer of data, between the database format to/from JSON format used in the frontend
- pages: Next page
- pages/api: API endpoints
- services: services for the backend (for API)
- theme: CSS theme
- utils: utilities functions necessary for frontend and backend

## 🖋️ Conventions for commit messages

You need to use the commit lint convention for commit message.

I.e, you must specify a type in prefix position, for the message using one of the following:

- build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
- ci: Changes to our CI configuration files and scripts (example scopes: Circle, BrowserStack, SauceLabs)
- docs: Documentation only changes
- feat: A new feature
- fix: A bug fix
- perf: A code change that improves performance
- refactor: A code change that neither fixes a bug nor adds a feature
- style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- test: Adding missing tests or correcting existing tests
- chore: for tedious and necessary work, not listed before (try to minimize the use of this umbrella term)

As a reference see https://github.com/conventional-changelog/commitlint/tree/master/@commitlint/config-angular.

## Conventions for closing automatically issues

Add in commit message "Closes #123" where 123 is the issues's id to close.

## 📈 Matomo

`trackEvent(category, action, [name], [value])`

avec :
- category/action:
    - acte
        - déclaration
        - suppression
        - modification
        - lecture
    - etp
        - déclaration
        - modification
        - validation annuelle
    - stat
        - lecture vivant
        - lecture thanato
        - lecture globale
    - authentification
        - connexion
        - déconnexion
        - erreur

## Annex

### Add askers reminder

```sql
INSERT INTO askers (name, type) VALUES ('CRS autoroutière', 'primary');
```

test 1
test 2


CHANGELOG.md
## [1.12.5](https://github.com/SocialGouv/medle/compare/v1.12.4...v1.12.5) (2021-09-15)


### Bug Fixes

* **ci:** upgrade gitlab-ci-yml@23.3.3 ([6dc754f](https://github.com/SocialGouv/medle/commit/6dc754f93f79c52a4d0191ab60cf8dbca4634071))

## [1.12.4](https://github.com/SocialGouv/medle/compare/v1.12.3...v1.12.4) (2021-09-15)


### Bug Fixes

* **ci:** use gitlab-ci-yml@23.3.2 ([03d1e99](https://github.com/SocialGouv/medle/commit/03d1e992a218335b35ef85758f5cd3bc63bb0e1c))

## [1.12.3](https://github.com/SocialGouv/medle/compare/v1.12.2...v1.12.3) (2021-09-15)


### Bug Fixes

* **ci:** upgrade gitlab-ci-yml ([#178](https://github.com/SocialGouv/medle/issues/178)) ([bc73c19](https://github.com/SocialGouv/medle/commit/bc73c19dd2f25e978370e5fb7a0e8a95daa279df))

## [1.12.2](https://github.com/SocialGouv/medle/compare/v1.12.1...v1.12.2) (2021-06-24)


### Bug Fixes

* **ci:** use new cluster for seeds ([#159](https://github.com/SocialGouv/medle/issues/159)) ([80d0c30](https://github.com/SocialGouv/medle/commit/80d0c30209a9291de50063bc1642f851795bc9df))

## [1.12.1](https://github.com/SocialGouv/medle/compare/v1.12.0...v1.12.1) (2021-06-16)


### Bug Fixes

* **ci:** update dependency socialgouv/gitlab-ci-yml to v23 ([#157](https://github.com/SocialGouv/medle/issues/157)) ([02edc42](https://github.com/SocialGouv/medle/commit/02edc422d1df7aeb7b1a13926a48eee406b1feb9))

# [1.12.0](https://github.com/SocialGouv/medle/compare/v1.11.5...v1.12.0) (2021-06-10)


### Features

* realease for attacks administration ([ded1b3b](https://github.com/SocialGouv/medle/commit/ded1b3bfc94ae31f4bbdfef35018274b574e5b20))

## [1.11.5](https://github.com/SocialGouv/medle/compare/v1.11.4...v1.11.5) (2021-06-07)


### Bug Fixes

* **export:** bug on export ([#154](https://github.com/SocialGouv/medle/issues/154)) ([6b55620](https://github.com/SocialGouv/medle/commit/6b55620c647ad381f91e6c22ab4c800155634f34))
* **k8s:** increate memory limit ([#153](https://github.com/SocialGouv/medle/issues/153)) ([166d1af](https://github.com/SocialGouv/medle/commit/166d1af63c026d32c582c1d5d287b7582de73ac5))

## [1.11.4](https://github.com/SocialGouv/medle/compare/v1.11.3...v1.11.4) (2021-06-07)


### Bug Fixes

* **release:** run release ([9a60953](https://github.com/SocialGouv/medle/commit/9a609539bbeb58ef3ff0efc90217b73c4b39f37e))
* decrease the number of exportables acts ([#152](https://github.com/SocialGouv/medle/issues/152)) ([fceb4b7](https://github.com/SocialGouv/medle/commit/fceb4b76bef475d45d4dfaf2008ed25f2a5b342d))

## [1.11.3](https://github.com/SocialGouv/medle/compare/v1.11.2...v1.11.3) (2021-06-07)


### Bug Fixes

* **k8s:** upgrade mem limits ([#151](https://github.com/SocialGouv/medle/issues/151)) ([4b4e15c](https://github.com/SocialGouv/medle/commit/4b4e15cebb85acad88f881f40f2fe76269ef75c3))

## [1.11.2](https://github.com/SocialGouv/medle/compare/v1.11.1...v1.11.2) (2021-05-27)


### Bug Fixes

* **ci:** another upgrade ([503ea43](https://github.com/SocialGouv/medle/commit/503ea43c8503f60f69d01c0933096efd985a09ff))

## [1.11.1](https://github.com/SocialGouv/medle/compare/v1.11.0...v1.11.1) (2021-05-26)


### Bug Fixes

* **ci:** upgrade deps ([#148](https://github.com/SocialGouv/medle/issues/148)) ([c44869d](https://github.com/SocialGouv/medle/commit/c44869d3a26ab5c8affd288c4a60fe6f43ceae04))

# [1.11.0](https://github.com/SocialGouv/medle/compare/v1.10.2...v1.11.0) (2021-05-05)


### Features

* add new profile ([#144](https://github.com/SocialGouv/medle/issues/144)) ([41bc96b](https://github.com/SocialGouv/medle/commit/41bc96b128e89252292c8a0a33973b53def6e243))

## [1.10.2](https://github.com/SocialGouv/medle/compare/v1.10.1...v1.10.2) (2021-04-22)


### Bug Fixes

* **gitlab:** update to SocialGouv/gitlab-ci-yml@v21.0.2 ([2c2cdcc](https://github.com/SocialGouv/medle/commit/2c2cdcc1ecdbd351ad59596c5f3d5535364d7161))

## [1.10.1](https://github.com/SocialGouv/medle/compare/v1.10.0...v1.10.1) (2021-04-20)


### Bug Fixes

* pb build time variables in k8s. ([#142](https://github.com/SocialGouv/medle/issues/142)) ([9f14fe8](https://github.com/SocialGouv/medle/commit/9f14fe8422c8a4d4e0776ae450efd3af2516cd9d))

# [1.10.0](https://github.com/SocialGouv/medle/compare/v1.9.1...v1.10.0) (2021-04-19)


### Features

* add forgot and reset password pages and api ([#140](https://github.com/SocialGouv/medle/issues/140)) ([402519d](https://github.com/SocialGouv/medle/commit/402519daf8ae112a151146861552461ac5a93495))

## [1.9.1](https://github.com/SocialGouv/medle/compare/v1.9.0...v1.9.1) (2021-04-14)


### Bug Fixes

* add tipimail secrets ([#141](https://github.com/SocialGouv/medle/issues/141)) ([027b993](https://github.com/SocialGouv/medle/commit/027b9933d86e0157a75c6cd0bcb32f6da5de9f50))

# [1.9.0](https://github.com/SocialGouv/medle/compare/v1.8.2...v1.9.0) (2021-04-12)


### Features

* update README with the use of Kubernetes now ([f448e1d](https://github.com/SocialGouv/medle/commit/f448e1da9e454f31c18577bc7a4ec4c9f5bb1bc4))

## [1.8.2](https://github.com/SocialGouv/medle/compare/v1.8.1...v1.8.2) (2021-04-09)


### Bug Fixes

* **k8s:** DB prod secrets ([#136](https://github.com/SocialGouv/medle/issues/136)) ([20b1b3e](https://github.com/SocialGouv/medle/commit/20b1b3e22a6078ee7d7c6f24aacbcf2e3f2f6708))
* **k8s:** restore real subdomain ([#137](https://github.com/SocialGouv/medle/issues/137)) ([951b123](https://github.com/SocialGouv/medle/commit/951b1237a5a71f42ac690645865214817690ce63))

## [1.8.1](https://github.com/SocialGouv/medle/compare/v1.8.0...v1.8.1) (2021-04-09)


### Bug Fixes

* error on etp export ([#138](https://github.com/SocialGouv/medle/issues/138)) ([67d45f4](https://github.com/SocialGouv/medle/commit/67d45f4552326563776dc59ed132dca4a923d4b7))

# [1.8.0](https://github.com/SocialGouv/medle/compare/v1.7.3...v1.8.0) (2021-03-22)


### Features


src/utils/date.js
import moment from "moment"
import getConfig from "next/config"

import { logError } from "../utils/logger"

const { publicRuntimeConfig } = getConfig() || {}

export const NAME_MONTHS = {
  "01": "janvier",
  "02": "février",
  "03": "mars",
  "04": "avril",
  "05": "mai",
  "06": "juin",
  "07": "juillet",
  "08": "août",
  "09": "septembre",
  10: "octobre",
  11: "novembre",
  12: "décembre",
}

export const FORMAT_DATE = "DD/MM/YYYY"
export const ISO_DATE = "YYYY-MM-DD"
export const ISO_TIME = "YYYY-MM-DDTHH:mm:ssZ"

export const now = () =>
  (publicRuntimeConfig?.TEST_CURRENT_DATE && moment(publicRuntimeConfig?.TEST_CURRENT_DATE, FORMAT_DATE)) || moment()

export const isValidIsoDate = (date) => date && moment(date, ISO_DATE, true).isValid()

export const frToIso = (date) => {
  const parts = date.split("/")

  if (parts.length !== 3) {
    logError("Problème de parsing de date")
    return date
  }
  const [day, month, year] = parts
  return `${year}-${month}-${day}`
}

export const isoToFr = (date) => {
  const parts = date.split("-")

  if (parts.length < 3) {
    logError("Problème de parsing de date")
    return date
  }

  const dayParts = parts[2].split("T") // Handle time part if present.

  const day = dayParts.length > 1 ? dayParts[0] : parts[2]

  const [year, month] = parts
  return `${day}/${month}/${year}`
}
